### PR TITLE
Feat/sample metric description

### LINKED ISSUE #24: SampleMetricDescription
Implement an abstract class for handling per-sample metric calculations.

Requires #1

### LINKED ISSUE #25: Alpha-Precision SampleMetricDescription
Implement a SampleMetricDescription for calculating alpha-precision on a per-sample basis.

Make sure to append to #29 if applicable.

Requires #24

### PR DESCRIPTION
**Pull request by sporreking**
Note that there were several additional features / commits, other than those mentioned by the issues.

**The conda environment must be rebuilt in order to access `scikit-learn`.**

Closes #24
Closes #25

### CODE REVIEW

File: src/metric/EvaluationEmbedding.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+
+def train(dataset: Dataset, start_state: ModelUtil.AuxModelInfo = None) -> None:
+    """
+    Trains the evaluation embedding model and saves it to disk for
+    different points along the training. The model is saved as an
+    auxiliary model using `util.ModelUtil`, under the name given
+    by the `metric.EvaluationEmbedding.AUX_MODEL_NAME` constant.
+
+    Args:
+        dataset (Dataset): The dataste to train on.
```
Comment (tediiieee) [inline src/metric/EvaluationEmbedding.py:277]: dataste -> dataset

File: src/metric/SampleMetricDescription.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+from typing import Any
+
+from src.dataset.Dataset import Dataset
+
+"""
+SampleMetricDescription(abstract) {
+  abstract name() -> str
+  abstract calc(data: (*)) -> array[metric]
+}
+"""
```
Comment (tediiieee) [inline src/metric/SampleMetricDescription.py:14]: old stuff?

File: src/metric/EvaluationEmbedding.py [hunk 3]
```diff
@@ -406,6 +410,90 @@ def train(dataset: Dataset, start_state: ModelUtil.AuxModelInfo = None) -> None:
...
 
 
+#! Projecting real samples
+
+
+def _ds_proj_file_name(dataset: Dataset) -> str:
+    return f"{DS_PROJ_FILE_PREFIX}_{dataset.get_name(dataset.get_resolution())}"
+
+
+def project(dataset: Dataset) -> None:
```
Comment (tediiieee) [inline src/metric/EvaluationEmbedding.py:420]: Couldnt project return the projection instead of None. With this change you could sometimes skip unecessary I/O through project() ->  p = get_projections() and instead just write p = project()?

File: src/metric/AlphaPrecisionSampleMetricDescription.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+class AlphaPrecisionSampleMetricDescription(SampleMetricDescription):
+    @staticmethod
+    def get_name() -> str:
+        return "AlphaPrecision"
+
+    @staticmethod
+    def setup(dataset: Dataset, mode: str = SETUP_MODE_CONTINUE) -> None:
+
+        # Check if valid mode
+        if mode not in (SETUP_MODE_RESTART, SETUP_MODE_CONTINUE, SETUP_MODE_PROJECT):
```
Comment (tediiieee) [inline src/metric/AlphaPrecisionSampleMetricDescription.py:34]: save collection of modes (as a constant) with the constants instead? Easier to add new modes that way IMO.

File: src/metric/AlphaPrecisionSampleMetricDescription.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                [
+                    T.ToTensor(),
+                    T.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
+                ]
+            ),
+        )
+
+        # Create population loader
+        pop_loader = torch.utils.data.DataLoader(
+            pop_images, batch_size=1, shuffle=False, num_workers=2
```
Comment (tediiieee) [inline src/metric/AlphaPrecisionSampleMetricDescription.py:100]: is batch_size = 1 reasonable? or just that low for testing?

**Review by tediiieee — CHANGES_REQUESTED** [review]
Looks good, if you want it to look freakin rad, check the small comments i made.

File: src/metric/AlphaPrecisionSampleMetricDescription.py (hunk 4 above)
Comment (sporreking) [reply src/metric/AlphaPrecisionSampleMetricDescription.py:34]: This will most likely be overridden / handled in #81.

Good catch however, this is indeed sub-optimal and your proposed solution is probably the best approach to the issue.

File: src/metric/AlphaPrecisionSampleMetricDescription.py (hunk 5 above)
Comment (sporreking) [reply src/metric/AlphaPrecisionSampleMetricDescription.py:100]: This was temporary, I forgot to add support for batch size

**Review by tediiieee — APPROVED** [review]
Nice.
